mabwiser/approximate.py

```diff
@@ -0,0 +1,170 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import defaultdict
+from copy import deepcopy
+from itertools import chain
+from typing import List, NoReturn, Optional, Union
+
+import numpy as np
+from joblib import Parallel, delayed
+
+from mabwiser.greedy import _EpsilonGreedy
+from mabwiser.linear import _Linear
+from mabwiser.neighbors import _Neighbors
+from mabwiser.popularity import _Popularity
+from mabwiser.rand import _Random
+from mabwiser.softmax import _Softmax
+from mabwiser.thompson import _ThompsonSampling
+from mabwiser.ucb import _UCB1
+from mabwiser.utils import Arm, _BaseRNG, create_rng
+
+
+class _ApproximateNeighbors(_Neighbors):
+
+    def fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+        # Initialize planes
+        self._initialize(contexts.shape[1])
+
+        # Set the historical data for prediction
+        self.decisions = decisions
+        self.contexts = contexts
+
+        # Binarize the rewards if using Thompson Sampling
+        if isinstance(self.lp, _ThompsonSampling) and self.lp.binarizer:
+            self.rewards = self._binarize_ts_rewards(decisions, rewards)
+        else:
+            self.rewards = rewards
+
+        # Fit hashes for each training context
+        self._fit_operation(contexts)
+
+    def partial_fit(self, decisions: np.ndarray, rewards: np.ndarray,
```

Comment:
This is a little confusing to me. In this default partial_fit implementation, are we effectively retraining the whole model?

Reply:
For LSH Nearest it just adds the new contexts to the existing hash tables. But now that it's been pointed out, I realize I forgot to update the indexes for those contexts. So thanks for noticing this!

If an algorithm needs to be retrained from scratch for each partial fit (e.g. necessary with the annoy library) it can access the updated self.contexts.